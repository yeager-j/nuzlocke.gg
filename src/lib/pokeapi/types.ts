/**
 * Represents a location within a game where an encounter can take place. `encounters` uses a Map and Set here to avoid
 * duplicate entries.
 *
 * This type is used to construct the location map of a game. It will be transformed into a JSON-compatible format (`PokemonGameLocation`).
 */
export type GameLocation = {
  name: string;
  encounters: Map<string, Set<string>>;
};

/**
 * The top-level structure that describes a Pok√©mon Game used in outputting JSON.
 */
export interface PokemonGame<T extends string> {
  id: string;
  name: string;
  locations: PokemonGameLocation<T>[];
}

/**
 * Represents a location within a game where an encounter can take place. Here, `encounters` is an object instead of a Map
 * so that it is compatible with JSON. This type is used in outputting a game's encounter JSON file.
 */
export interface PokemonGameLocation<T extends string> {
  id: T;
  name: string;
  encounters: {
    [method: string]: string[];
  };
}

/**
 * This type describes a map of locations to transformer functions. This type allows locations to have their data overriden
 * to handle edge-cases, such as Starters.
 */
export type LocationTransformer<T extends string> = Partial<
  Record<T, () => PokemonGameLocation<T>>
>;
