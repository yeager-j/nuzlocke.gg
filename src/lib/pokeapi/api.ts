import Pokedex from "pokedex-promise-v2";

export const pokeapi = new Pokedex();
