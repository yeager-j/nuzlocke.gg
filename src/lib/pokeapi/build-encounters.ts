import fs from "fs/promises";
import path from "path";

import {
  GameLocation,
  LocationTransformer,
  PokemonGame,
  PokemonGameLocation,
} from "@/lib/pokeapi/types";
import {
  applyLocationTransformation,
  getEncounterLocationsForGame,
  getEncountersAsObject,
  getJSONOutputPath,
} from "@/lib/pokeapi/utils";

/**
 * Assembles game data for a Pokémon game by processing location data and applying a custom transformer function.
 *
 * @param game An object containing the id and name of the Pokémon game.
 * @param locationOrder An array of location identifiers specifying the order of locations.
 * @param gameLocations A map of location ids to their corresponding GameLocation objects.
 * @param transformer A function that accepts the gameLocations map and returns a LocationTransformer specific to the game.
 * @return A Promise that resolves to a complete Pokémon game object, including all processed locations.
 */
export async function buildEncounterGameData<T extends string>(
  game: Pick<PokemonGame<T>, "id" | "name">,
  locationOrder: readonly T[],
  gameLocations: Map<string, GameLocation>,
  transformer: (
    gameLocations: Map<string, GameLocation>,
  ) => LocationTransformer<T>,
): Promise<PokemonGame<T>> {
  const locationHandlers = transformer(gameLocations);

  return {
    ...game,
    locations: locationOrder.reduce((acc, prev) => {
      if (prev in locationHandlers) {
        const transformedLocation = applyLocationTransformation(
          prev,
          locationHandlers,
        );

        return [...acc, transformedLocation];
      }

      const locationData = gameLocations.get(prev);

      if (!locationData) {
        throw new Error(`Unable to find location: ${prev}`);
      }

      const pokemonLocation = {
        id: prev,
        name: locationData.name,
        encounters: getEncountersAsObject(locationData.encounters),
      };

      return [...acc, pokemonLocation];
    }, [] as PokemonGameLocation<T>[]),
  };
}

/**
 * Writes a JSON file for a specific Pokémon game with its processed location data.
 *
 * @param {Pick<PokemonGame<T>, "id" | "name">} game - An object containing the game's `id` and `name`.
 * @param {readonly T[]} locationOrder - An ordered list of location identifiers used to structure the game locations.
 * @param {(gameLocations: Map<string, GameLocation>) => LocationTransformer<T>} transformer - A function that transforms the game locations into a specific format.
 */
export async function writeEncounterJSONFile<T extends string>(
  game: Pick<PokemonGame<T>, "id" | "name">,
  locationOrder: readonly T[],
  transformer: (
    gameLocations: Map<string, GameLocation>,
  ) => LocationTransformer<T>,
) {
  const gameLocations = await getEncounterLocationsForGame(game.id);

  const gameData = await buildEncounterGameData(
    game,
    locationOrder,
    gameLocations,
    transformer,
  );

  await fs.mkdir(getJSONOutputPath(game.id), {
    recursive: true,
  });

  await fs.writeFile(
    path.join(getJSONOutputPath(game.id), `encounters.json`),
    JSON.stringify(gameData, null, 2),
  );
}
