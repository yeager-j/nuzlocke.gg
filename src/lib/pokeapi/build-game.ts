import fs from "fs/promises";
import path from "path";

import {
  GameLocation,
  LocationTransformer,
  PokemonGame,
  PokemonLocation,
} from "@/lib/pokeapi/types";
import {
  applyLocationTransformation,
  getEncounterLocationsForGame,
  getEncountersAsObject,
  LOCATION_OUTPUT_PATH,
} from "@/lib/pokeapi/utils";

/**
 * Assembles game data for a Pokémon game by processing location data and applying a custom transformer function.
 *
 * @param game An object containing the id and name of the Pokémon game.
 * @param locationOrder An array of location identifiers specifying the order of locations.
 * @param gameLocations A map of location ids to their corresponding GameLocation objects.
 * @param transformer A function that accepts the gameLocations map and returns a LocationTransformer specific to the game.
 * @return A Promise that resolves to a complete Pokémon game object, including all processed locations.
 */
export async function buildGameData<T extends string>(
  game: Pick<PokemonGame<T>, "id" | "name">,
  locationOrder: readonly T[],
  gameLocations: Map<string, GameLocation>,
  transformer: (
    gameLocations: Map<string, GameLocation>,
  ) => LocationTransformer<T>,
): Promise<PokemonGame<T>> {
  const locationHandlers = transformer(gameLocations);

  return {
    ...game,
    locations: locationOrder.reduce((acc, prev) => {
      if (prev in locationHandlers) {
        const transformedLocation = applyLocationTransformation(
          prev,
          locationHandlers,
        );

        return [...acc, transformedLocation];
      }

      const locationData = gameLocations.get(prev);

      if (!locationData) {
        throw new Error(`Unable to find location: ${prev}`);
      }

      const pokemonLocation = {
        id: prev,
        name: locationData.name,
        encounters: getEncountersAsObject(locationData.encounters),
      };

      return [...acc, pokemonLocation];
    }, [] as PokemonLocation<T>[]),
  };
}

/**
 * Writes a JSON file for a specific Pokémon game with its processed location data.
 *
 * @param {Pick<PokemonGame<T>, "id" | "name">} game - An object containing the game's `id` and `name`.
 * @param {readonly T[]} locationOrder - An ordered list of location identifiers used to structure the game locations.
 * @param {(gameLocations: Map<string, GameLocation>) => LocationTransformer<T>} transformer - A function that transforms the game locations into a specific format.
 */
export async function writeJSONFile<T extends string>(
  game: Pick<PokemonGame<T>, "id" | "name">,
  locationOrder: readonly T[],
  transformer: (
    gameLocations: Map<string, GameLocation>,
  ) => LocationTransformer<T>,
) {
  const gameLocations = await getEncounterLocationsForGame(game.id);

  const gameData = await buildGameData(
    game,
    locationOrder,
    gameLocations,
    transformer,
  );

  await fs.mkdir(LOCATION_OUTPUT_PATH, {
    recursive: true,
  });

  await fs.writeFile(
    path.join(LOCATION_OUTPUT_PATH, `${game.id}.json`),
    JSON.stringify(gameData, null, 2),
  );
}
