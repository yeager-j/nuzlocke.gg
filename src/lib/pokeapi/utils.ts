import path from "path";
import { Name } from "pokedex-promise-v2";

import { pokeapi } from "@/lib/pokeapi/api";

export type GameLocation = {
  name: string;
  encounters: Map<string, Set<string>>;
};

export const LOCATION_OUTPUT_PATH = path.join(process.cwd(), "public/games");

/**
 * Converts the encounters map of a game location into a plain object.
 *
 * @param {Map} encounters - A Map where keys are encounter types and values are Sets of encounter details.
 * @return {Object} A plain object where keys match the encounter types and values are arrays of encounter details.
 */
export function getEncountersAsObject(encounters: GameLocation["encounters"]): {
  [K: string]: string[];
} {
  return Object.fromEntries(
    Array.from(encounters.entries()).map(([key, valueSet]) => [
      key,
      Array.from(valueSet),
    ]),
  );
}

/**
 * Retrieves the English name from an array of name objects.
 *
 * @param {Name[]} names - An array of name objects, each containing a language property and a name property.
 * @return {string} The English name found in the input array.
 * @throws {Error} If no English name is found in the input array.
 */
export function getEnglishName(names: Name[]): string {
  const engName = names.find((n) => n.language.name === "en");

  if (!engName) {
    throw new Error("No english name found");
  }

  return engName.name;
}

/**
 * Retrieves encounter locations and details for a specified game.
 *
 * @param {string} game - The name of the game for which to retrieve encounter locations, e.g. "red", "platinum", "ultra-sun".
 * @return {Promise<Map<string, GameLocation>>} A Promise that resolves to a map of location names
 * to detailed encounter locations, including encounter methods and Pok√©mon details.
 */
export async function getEncounterLocationsForGame(
  game: string,
): Promise<Map<string, GameLocation>> {
  const version = await pokeapi.getVersionByName(game);

  const versionGroup = await pokeapi.getVersionGroupByName(
    version.version_group.name,
  );

  const regions = await pokeapi.getRegionByName(
    versionGroup.regions.map((r) => r.name),
  );

  const locations = await Promise.all(
    regions
      .flatMap((r) => r.locations)
      .map((l) => pokeapi.getLocationByName(l.name)),
  );

  const gameLocations = new Map(
    locations.map((location) => [
      location.name,
      {
        name: getEnglishName(location.names),
        encounters: new Map(),
      },
    ]),
  );

  const locationAreas = await Promise.all(
    locations
      .flatMap((l) => l.areas)
      .map((a) => pokeapi.getLocationAreaByName(a.name)),
  );

  locationAreas.forEach((area) => {
    const location = gameLocations.get(area.location.name);
    if (!location) return;

    area.pokemon_encounters
      .flatMap((encounter) => {
        const gameVersion = encounter.version_details.find(
          (v) => v.version.name === game,
        );
        if (!gameVersion) return [];

        const pokemonName = encounter.pokemon.name;

        return gameVersion.encounter_details.map((detail) => ({
          method: detail.method.name,
          pokemon: pokemonName,
        }));
      })
      .forEach(({ method, pokemon }) => {
        const pokemonSet = location.encounters.get(method) || new Set();
        pokemonSet.add(pokemon);
        location.encounters.set(method, pokemonSet);
      });
  });

  return gameLocations;
}
