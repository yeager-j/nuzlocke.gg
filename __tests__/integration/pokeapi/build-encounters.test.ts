import path from "path";
import { vol } from "memfs";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";

import { setupMockedPokeAPI } from "../../pokeapi-test-utils";

import { writeEncounterJSONFile } from "@/lib/pokeapi/build-encounters";
import { LocationTransformer, PokemonGame } from "@/lib/pokeapi/types";
import { getJSONOutputPath } from "@/lib/pokeapi/utils";

vi.mock("@/lib/pokeapi/api", () => {
  return {
    pokeapi: {
      getVersionByName: vi.fn(),
      getVersionGroupByName: vi.fn(),
      getRegionByName: vi.fn(),
      getLocationByName: vi.fn(),
      getLocationAreaByName: vi.fn(),
    },
  };
});

vi.mock("fs/promises");

describe("writeJSONFile", async () => {
  beforeEach(async () => {
    vol.reset();

    await setupMockedPokeAPI();
  });

  afterEach(() => {
    vi.resetAllMocks();
  });

  it("should write a JSON file with the correct game data", async () => {
    const locationOrder = [
      "pallet-town",
      "kanto-route-1",
      "viridian-city",
      "viridian-forest",
    ] as const;

    type TestLocation = (typeof locationOrder)[number];

    const game: Pick<PokemonGame<TestLocation>, "id" | "name"> = {
      id: "red",
      name: "Pokémon Red",
    };

    const transformer = (): LocationTransformer<TestLocation> => {
      return {};
    };

    await writeEncounterJSONFile(game, locationOrder, transformer);

    const gameData = vol.readFileSync(
      path.join(getJSONOutputPath("red"), "encounters.json"),
      "utf-8",
    );

    expect(gameData.toString()).toMatchSnapshot();
  });

  it("should apply custom location transformers when provided", async () => {
    const locationOrder = [
      "pallet-town",
      "kanto-route-1",
      "viridian-city",
      "viridian-forest",
    ] as const;

    type TestLocation = (typeof locationOrder)[number];

    const game: Pick<PokemonGame<TestLocation>, "id" | "name"> = {
      id: "red",
      name: "Pokémon Red",
    };

    const transformer = (): LocationTransformer<TestLocation> => {
      return {
        "viridian-forest": () => ({
          id: "viridian-forest",
          name: "Viridian Forest",
          encounters: {
            walk: ["pikachu", "caterpie", "weedle", "kakuna", "metapod"],
            special: ["bulbasaur"],
          },
        }),
      };
    };

    await writeEncounterJSONFile(game, locationOrder, transformer);

    const gameData = vol.readFileSync(
      path.join(getJSONOutputPath("red"), "encounters.json"),
      "utf-8",
    );

    expect(gameData.toString()).toMatchSnapshot();
  });
});
