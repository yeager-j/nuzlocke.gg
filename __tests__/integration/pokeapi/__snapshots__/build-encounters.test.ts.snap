// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`writeJSONFile > should apply custom location transformers when provided 1`] = `
"{
  "id": "red",
  "name": "Pokémon Red",
  "locations": [
    {
      "id": "pallet-town",
      "name": "Pallet Town",
      "encounters": {
        "gift": [
          "bulbasaur",
          "charmander",
          "squirtle"
        ],
        "good-rod": [
          "poliwag",
          "goldeen"
        ],
        "super-rod": [
          "poliwag",
          "tentacool"
        ],
        "old-rod": [
          "magikarp"
        ]
      }
    },
    {
      "id": "kanto-route-1",
      "name": "Route 1",
      "encounters": {
        "walk": [
          "pidgey",
          "rattata"
        ]
      }
    },
    {
      "id": "viridian-city",
      "name": "Viridian City",
      "encounters": {
        "good-rod": [
          "poliwag",
          "goldeen"
        ],
        "super-rod": [
          "poliwag",
          "tentacool"
        ],
        "old-rod": [
          "magikarp"
        ]
      }
    },
    {
      "id": "viridian-forest",
      "name": "Viridian Forest",
      "encounters": {
        "walk": [
          "pikachu",
          "caterpie",
          "weedle",
          "kakuna",
          "metapod"
        ],
        "special": [
          "bulbasaur"
        ]
      }
    }
  ]
}"
`;

exports[`writeJSONFile > should write a JSON file with the correct game data 1`] = `
"{
  "id": "red",
  "name": "Pokémon Red",
  "locations": [
    {
      "id": "pallet-town",
      "name": "Pallet Town",
      "encounters": {
        "gift": [
          "bulbasaur",
          "charmander",
          "squirtle"
        ],
        "good-rod": [
          "poliwag",
          "goldeen"
        ],
        "super-rod": [
          "poliwag",
          "tentacool"
        ],
        "old-rod": [
          "magikarp"
        ]
      }
    },
    {
      "id": "kanto-route-1",
      "name": "Route 1",
      "encounters": {
        "walk": [
          "pidgey",
          "rattata"
        ]
      }
    },
    {
      "id": "viridian-city",
      "name": "Viridian City",
      "encounters": {
        "good-rod": [
          "poliwag",
          "goldeen"
        ],
        "super-rod": [
          "poliwag",
          "tentacool"
        ],
        "old-rod": [
          "magikarp"
        ]
      }
    },
    {
      "id": "viridian-forest",
      "name": "Viridian Forest",
      "encounters": {
        "walk": [
          "caterpie",
          "metapod",
          "weedle",
          "kakuna",
          "pikachu"
        ]
      }
    }
  ]
}"
`;
